### ISSUE TITLE
`display-name` `ignoreTranspilerName` is too strict

### ISSUE THREAD

**Issue by ljharb** [yannickcr/eslint-plugin-react#1165]
The purpose of `ignoreTranspilerName` is to ignore inferred names:
```jsx
const SFC = () => <div />; // inferred name
const Foo = class extends React.Component {} // inferred name
```
It should not, however, ignore explicit names:
```jsx
function SFC() { return <div />; } // explicit name
class Foo extends React.Component {} // explicit name
```

I think that the `ignoreTranspilerName` option should be deprecated, and replaced with new options that cover the current use cases:
 1. always require an explicit `displayName` (`ignoreTranspilerName: true`)
 1. only require an explicit `displayName` on `React.createClass`es (`ignoreTranspilerName: false`)

As well as cover these additional use cases:
 1. always require an explicit *name* - ie, an explicit displayName *or* a named function or named `class
 1. always require a name - either explicit, or ES6-inferred
 1. Never allow an explicit `displayName`? (i don't really care about this one, but it seems useful for consistency)

**Comment by ljharb** [comment]
@yannickcr @lencioni @EvNaverniouk thoughts?

**Comment by rodoabad** [comment]
Voting for option 2!

**Comment by ljharb** [comment]
These aren't exclusive options, they're use cases :-) also there's two lists, which "option 2" do you have in mind?

**Comment by rodoabad** [comment]
@ljharb if it can be inferred then I think not explicitly setting a `displayName` is fine. Anonymous functions or classes though should still be required to have a display name.

```js
export default function () {
  return <span>{'foo'}</span>;
}
```

This should throw an error.

**Comment by ljharb** [comment]
Yes, that's what it's already doing - but in my example, I need "always require an explicit name", because I never want to rely on name inference.

**Comment by rodoabad** [comment]
@ljharb yeah you're talking about having three options right? I agree to that.

1. Explicit everything.
2. Inferred if possible otherwise you gotta do it bro!
3. Do not set a display name.

This is the one that I like that's why I said I like option 2. Now I just realized it could mean "2." Hahaha.

**Comment by golopot** [comment]
How about `"always" | "only-unnamed" | "only-create-class"`?
```
┌────────────────────────────────┬────────┬──────────────┬────────────────────┐
│                                │ always │ only-unnamed │  only-create-class │
├────────────────────────────────┼────────┼──────────────┼────────────────────┤
│ function Cat() {}              │ fail   │  pass        │  pass              │
│ class Cat extends Component {} │ fail   │  pass        │  pass              │
│ export default () => <></>     │ fail   │  fail        │  pass              │
│ exports.Cat = () => <></>      │ fail   │  fail        │  pass              │
│ const Cat = createClass({})    │ fail   │  fail        │  fail              │
└────────────────────────────────┴────────┴──────────────┴────────────────────┘
```

**Comment by ljharb** [comment]
@golopot `exports.Cat = () => {}` doesn't have inference; maybe add `const Cat = () => {}` which should be `fail`/`fail`/`pass`?

**Comment by SevenOutman** [comment]
Any update on this? Agree that it should not report positive on SFCs that have explicit names - which is almost every component in 2022.

```jsx
function DontComplainAboutMyName() {
  return <div></div>;
}
```

**Comment by ljharb** [comment]
@SevenOutman it doesn’t report on those, as far as I’m aware. This issue is about reporting on arrow function components that lack inferred names.

**Comment by SevenOutman** [comment]
> it doesn’t report on those, as far as I’m aware.

Oops, turns out it was reporting anonymous arrow function components wrapped with `forwardRef`. My mistake.
